### PR TITLE
Openapi gen

### PR DESCRIPTION
**Pull request by c-harding**
Surprised this is only 152 lines, was worried it was going to be bigger.

- Creates a function to generate the OpenAPI description for an error response type.
- Adds a library for Yaml-ifying it, as used in a test.
- Adds a trait for Jsonable structures with a custom Json representation
- I got very confused by the implicit nature of the `circe` library, but I think most of weirdest magic is factored out now. Now all that is needed is `import io.circe.syntax._` to get `.asJson` and `import io.circe.yaml.syntax.AsYaml` to get `.asYaml`. However, `temple.generate.target.openapi.Jsonable#encodeJsonable`/`temple.generate.target.openapi.Jsonable.encodeJsonable` is duplicated as protected to save on needing to import it elsewhere inside concrete `Jsonable` implementations.

Side tidying:

- Fixes a warning of some redundant parens in `build.sbt`

### CODE REVIEW

File: src/main/scala/temple/generate/target/openapi/Jsonable.scala [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package temple.generate.target.openapi
+
+import io.circe.{Encoder, Json}
+
+private[openapi] trait Jsonable {
```
Comment (jaylees14) [inline src/main/scala/temple/generate/target/openapi/Jsonable.scala]: How about `JsonEncodable`?

File: src/main/scala/temple/generate/target/openapi/Jsonable.scala [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+  // Required so that nested Jsonable interfaces always call the correct nested version
+  implicit final protected def encodeJsonable[T <: Jsonable]: Encoder[T] = Jsonable.encodeJsonable
+}
+
+private[openapi] object Jsonable {
+
+  /** Create an encoder for JSON objects by providing a function to map them to options of values */
+  private def mapSequenceEncoder[T](f: T => Map[String, Option[Json]]): Encoder[T] =
+    (a: T) => Json.obj(f(a).iterator.flatMap { case (str, maybeJson) => maybeJson.map(str -> _) }.toSeq: _*)
```
Comment (jaylees14) [inline src/main/scala/temple/generate/target/openapi/Jsonable.scala]: Can we break this line up with some variable names? It's pretty hard to follow

File: src/main/scala/temple/generate/target/openapi/OpenAPIType.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import scala.collection.immutable.ListMap
+
+sealed abstract private[openapi] class OpenAPIType(fieldEntries: Seq[(String, Json)]) extends Jsonable {
+  def typeString: String
+  final lazy val customFields: Map[String, Json] = fieldEntries.to(ListMap)
+
+  override def toJsonMap: Map[String, Option[Json]] =
+    ListMap(
+      "type"       -> Some(typeString.asJson),
+      "properties" -> this.matchPartial { case openAPIType: OpenAPIObject => openAPIType.properties.asJson },
```
Comment (jaylees14) [inline src/main/scala/temple/generate/target/openapi/OpenAPIType.scala]: Just a thought - is this good design, for the abstract "parent" type to know about the "children" types?

File: src/main/scala/temple/generate/target/openapi/Response.scala [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+import io.circe.{Encoder, Json}
+
+import scala.collection.immutable.ListMap
+
+/** https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#responseObject */
+private[openapi] case class Response(
+  description: String,
+  content: Map[String, Response.MediaTypeObject],
+  required: Option[Boolean] = None,
```
Comment (jaylees14) [inline src/main/scala/temple/generate/target/openapi/Response.scala:13]: Kinda annoying that you'd have to wrap it in `Some` to use it, but I see why

**Review by jaylees14 — COMMENTED** [review]
Looks good, just a few general comments. Happy to approve if you don't agree tho 👍

File: src/main/scala/temple/generate/target/openapi/OpenAPIGenerator.scala [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+import scala.collection.immutable.ListMap
+
+object OpenAPIGenerator {
+
+  /** Create a Response representation for an error */
+  private[openapi] def generateError(description: String, code: Int, example: String): Response =
+    Response(
+      description,
+      ListMap(
```
Comment (Lewiky) [inline src/main/scala/temple/generate/target/openapi/OpenAPIGenerator.scala:14]: Did we work out that ListMap maintains dict order?

File: src/main/scala/temple/generate/target/openapi/OpenAPIType.scala [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      "items"      -> this.matchPartial { case openAPIType: OpenAPIArray => openAPIType.items.asJson },
```
Comment (Lewiky) [inline src/main/scala/temple/generate/target/openapi/OpenAPIType.scala]: matchPartial?

File: src/main/scala/temple/generate/target/openapi/OpenAPIType.scala [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    ) ++ customFields.view.mapValues(Some(_))
+}
+
+private[openapi] object OpenAPIType {
+
+  sealed case class OpenAPISimpleType(typeString: String, fieldEntries: (String, Json)*)
+      extends OpenAPIType(fieldEntries)
```
Comment (Lewiky) [inline src/main/scala/temple/generate/target/openapi/OpenAPIType.scala]: What is `typeString` in this case where it isn't defined? How does the `toJsonMap` method work in this case?

File: src/main/scala/temple/utils/MonadUtils.scala [hunk 8]
```diff
@@ -8,4 +8,11 @@ object MonadUtils {
     /** Reduces the Either to an item of the right type, by transforming a Left item if present */
     def fromEither(f: A => B): B = either.fold(f, identity)
   }
+
+  /** Adds the method `matchPartial` to any object, to call a partial function directly on it */
+  implicit class MatchPartial[A](x: A) {
```
Comment (Lewiky) [inline src/main/scala/temple/utils/MonadUtils.scala]: Ok so this is how it's defined, resolves my earlier comment but could we get a comment somewhere to point to MonadUtils?

File: src/main/scala/temple/generate/target/openapi/OpenAPIGenerator.scala [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+import temple.generate.target.openapi.OpenAPIType._
+import io.circe.syntax._
+
+import scala.collection.immutable.ListMap
+
+object OpenAPIGenerator {
+
+  /** Create a Response representation for an error */
+  private[openapi] def generateError(description: String, code: Int, example: String): Response =
```
Comment (Lewiky) [inline src/main/scala/temple/generate/target/openapi/OpenAPIGenerator.scala]: Where does `code` get used?

File: src/main/scala/temple/generate/target/openapi/OpenAPIGenerator.scala (hunk 5 above)
Comment (c-harding) [reply src/main/scala/temple/generate/target/openapi/OpenAPIGenerator.scala:14]: Yes

File: src/main/scala/temple/generate/target/openapi/OpenAPIType.scala (hunk 3 above)
Comment (c-harding) [reply src/main/scala/temple/generate/target/openapi/OpenAPIType.scala]: How does this alternative look? Removes the need for matchPartial, as each child passes up the additional fields they define
https://github.com/TempleEight/temple/compare/openapi-gen...openapi-gen-rewrite-openapitype-tojsonmap-2

Another alternative, but personally I think this seems a little more repetitive:
https://github.com/TempleEight/temple/compare/openapi-gen...openapi-gen-rewrite-openapitype-tojsonmap

File: src/main/scala/temple/generate/target/openapi/OpenAPIType.scala (hunk 7 above)
Comment (c-harding) [reply src/main/scala/temple/generate/target/openapi/OpenAPIType.scala]: Anything like number or string, e.g. in

https://github.com/TempleEight/temple/blob/5c637f8d7ed6ae7d72d0acfe30891632e924958f/src/main/scala/temple/generate/target/openapi/OpenAPIGenerator.scala#L16

File: src/main/scala/temple/generate/target/openapi/Response.scala (hunk 4 above)
Comment (c-harding) [reply src/main/scala/temple/generate/target/openapi/Response.scala:13]: Can you think of an alternative? Guess I could default to `false`, and only render it if `true`, but you may wish to be more explicit in some cases

File: src/main/scala/temple/generate/target/openapi/OpenAPIType.scala (hunk 7 above)
Comment (c-harding) [reply src/main/scala/temple/generate/target/openapi/OpenAPIType.scala]: In terms of how it works, it’s been updated slightly, but originally the definition of a value in a case class overrides a def in the parent class

**Review by jaylees14 — APPROVED** [review]

**Review by Lewiky — APPROVED** [review]
